torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -691,7 +692,23 @@ bool ProcessGroupNCCL::WorkNCCL::checkTimeout(
 
   LOG(ERROR) << exceptionMsg;
 
-  // Get the stack trace of the work at call time
+  std::exception_ptr exception_ptr =
+      std::make_exception_ptr(C10_BUILD_ERROR(DistBackendError, exceptionMsg));
+  if (!exception()) {
+    // if there is already an error, we don't override it
+    setException(exception_ptr);
+  }
+
+  // Mark future result as TIMEOUT
+  if (futureWorkResult_ && !futureWorkResult_->completed()) {
```

Comment:
LGTM that we moved this to checktimeout, but one caveat is that checkTimeout is being called not just from watchdog thread, but user thread too (e.g., when user calls work.wait(timeout)), so the callback registered by the user could be executed in the user thread in this case. But I think this should be OK, or even desirable.

Reply:
I think it is supported to mark completed *before* a callback is attached to a Future. The callback will directly get executed at attach time in that case.